lib/ansible/modules/database/mysql/mysql_db.py

```diff
@@ -346,81 +378,68 @@ def main():
-                changed = True
-            else:
-                try:
-                    changed = db_create(cursor, db, encoding, collation)
-                except Exception as e:
-                    module.fail_json(msg="error creating database: %s" % to_native(e),
-                                     exception=traceback.format_exc())
-            module.exit_json(changed=changed, db=db_name)
-
-        elif state == "import":
-            if module.check_mode:
-                module.exit_json(changed=True, db=db_name)
+    if not all_databases:
+        for each_database in db:
+            if db_exists(cursor, [each_database]):
+                existence_list.append(each_database)
             else:
-                try:
-                    changed = db_create(cursor, db, encoding, collation)
-                    if changed:
-                        rc, stdout, stderr = db_import(module, login_host, login_user,
-                                                       login_password, db, target, all_databases,
-                                                       login_port, config_file, socket, ssl_cert, ssl_key, ssl_ca)
-                        if rc != 0:
-                            module.fail_json(msg="%s" % stderr)
-                        else:
-                            module.exit_json(changed=True, db=db_name, msg=stdout)
-                except Exception as e:
-                    module.fail_json(msg="error creating database: %s" % to_native(e),
-                                     exception=traceback.format_exc())
-
-        elif state == "absent":
-            module.exit_json(changed=False, db=db_name)
-
-        elif state == "dump":
-            if module.check_mode:
-                module.exit_json(changed=False, db=db_name)
-            module.fail_json(msg="Cannot dump database %r - not found" % (db_name))
+                non_existence_list.append(each_database)
+
+    if state == "absent":
+        if module.check_mode:
+            module.exit_json(changed=bool(existence_list), db=db_name, db_list=db)
+        try:
+            changed = db_delete(cursor, existence_list)
+        except Exception as e:
+            module.fail_json(msg="error deleting database: %s" % to_native(e))
+        module.exit_json(changed=changed, db=db_name, db_list=db)
+    elif state == "present":
+        if module.check_mode:
+            module.exit_json(changed=bool(non_existence_list), db=db_name, db_list=db)
+        changed = False
+        if non_existence_list:
+            try:
+                changed = db_create(cursor, non_existence_list, encoding, collation)
+            except Exception as e:
+                module.fail_json(msg="error creating database: %s" % to_native(e),
+                                 exception=traceback.format_exc())
+            # If atleast one database was created by db_create then changed should be True
+            if not changed and any_db_exists(non_existence_list):
```

Comment:
I don't think this works. In case `db_create` returns `False`, it can be that some of the DBs were created, or none (because they were already created between the existence check and the CREATE DATABASE command). In case `db_create` didn't fail and `non_existence_list` is not empty, `any_db_exists(non_existence_list)` will *always* return `True`. But `changed` might still have to be `False`.

Reply:
fixed with the suggestion provided in comment https://github.com/ansible/ansible/pull/58602#discussion_r302730546